python/dolfin/function/expression.py

```diff
@@ -1,47 +1,79 @@
 # -*- coding: utf-8 -*-
-# Copyright (C) 2018 Michal Habera
+# Copyright (C) 2018 Michal Habera, Jack S. Hale
 #
 # This file is part of DOLFIN (https://www.fenicsproject.org)
 #
 # SPDX-License-Identifier:    LGPL-3.0-or-later
+from typing import Callable
 
 import numba
 from petsc4py import PETSc
 
 from dolfin import cpp
 
 
-def numba_eval(func):
-    """Decorator for JITable eval method
+def numba_eval(*args, numba_jit_options: dict = {"nopython": True, "cache": True},
```

Comment:
I don't think it will work if the function can't be compiled ('nopython' not true), so maybe cache is the only 'real' option?

Reply:
Not sure here. More important is the `cfunc` decorator. 
I think numba JIT could also work without nopython, but is super slow.

Reply:
I tweaked a test at 02c202b that shows that everything works happily with numba's (very slow) cfunc and jit object compilation mode.

https://numba.pydata.org/numba-doc/latest/glossary.html#term-object-mode

Reply:
Ignore the above, talking total rubbish. I won't protect the (advanced) user from overriding nopython and/or setting forceobj. Numba throws an exception, which is the pythonic behaviour. Test in place with xfail on exception `numba.errors.TypingError`.